### PR TITLE
Draft: Chat functions with live syncing

### PR DESCRIPTION
**Pull request by Bawnorton**
## Scope:
- [x] Message Editing
- [x] Message Pinning
- [x] Message Replying
- [ ] ~~Mentioning~~
- [x] Attaching media
- [x] Typing indicators

All features with live syncing so for example if client A is looking at message B in the pins view and client C edits message B, client A will see message B in the pins update live.

### CODE REVIEW

File: App/Components/Chat/CompChatView.razor [hunk 1]
```diff
@@ -168,73 +269,152 @@
...
     private Chat? Chat { get; set; }
+    private string ChatName { get; set; }
+    private bool ChatLoaded { get; set; }
+
+    private User? _otherUser;
+
+    /// <summary>
+    ///     Null if the chat is a group chat
+    /// </summary>
+    private User? OtherUser
```
Comment (bhunt02) [inline App/Components/Chat/CompChatView.razor:289]: noticing an issue here; this property doesn't adequately account for changes to the other user in the dm when switching DM views (i.e., the chat id parameter is set) the header remains the same, but the member list changes.

File: App/Components/Chat/CompChatView.razor [hunk 2]
```diff
@@ -168,73 +269,152 @@
...
     private bool IsMod { get; set; }
-    private List<LinkedMessage> Messages { get; set; } = [];
-    
-    private DateTime? _lastMessageSentAt;
-    
-    private string MessageInput { get; set; } = default!;
 
-    private string _hubUrl = default!;
-    private HubConnection? _hubConnection;
+    private List<MessageEntry> Messages { get; set; } = [];
```
Comment (bhunt02) [inline App/Components/Chat/CompChatView.razor:305]: Messages, EphemeralMessages & MessageIdToIndex are not cleared out when the chat is changed, so the same messages are taken to each different chat display :(

File: App/Classes/Services/UserServices/UserInteractionService.cs [hunk 3]
```diff
@@ -366,22 +366,31 @@ public async Task<List<User>> GetAllBlockedBy()
     }
 
     public List<IInteractionSubscriber> Subscribers { get; } = [];
+    private SemaphoreSlim SubscribersLock { get; } = new(1, 1);
```
Comment (bhunt02) [inline App/Classes/Services/UserServices/UserInteractionService.cs:369]: the semaphore here doesn't appear to work in preventing errors when the subscriber list is modified - can see that an error is thrown when unfriending someone while in the dm view for them

**Review by bhunt02 — CHANGES_REQUESTED** [review]
gave everything a thorough test as part of my work in #143, noticing some issues. i think this might be another instance of blazor being lazy with parameters

File: App/Components/Chat/CompChatView.razor [hunk 4]
```diff
@@ -250,113 +430,203 @@
...
+                .FirstAsync(m => m.ID == messageId);
+            CreateMessageEntry(message);
+            await JumpToMessage(message, false);
+        }
     }
 
     private async Task SendMessage(string messageText)
     {
-        if (string.IsNullOrWhiteSpace(messageText)) return;
+        if (string.IsNullOrWhiteSpace(messageText) && FileData.Count == 0) return;
```
[the remark this answers is not in the document]
Comment (Bawnorton) [reply App/Components/Chat/CompChatView.razor:628]: Is this no longer needed per #143? Should I include it anyway?

[the remark this answers is not in the document]
Comment (bhunt02) [reply App/Components/Chat/CompChatView.razor:616]: include it anyway for now, i am exploring into making a rudimentary basic 'system message' sort of interaction.

[the remark this answers is not in the document]
Comment (Bawnorton) [reply App/Components/Chat/CompChatView.razor:621]: alr, can probably use that for pinning too provided it's plug-and-play-able

**Review by bhunt02 — APPROVED** [review]

**Pull request merged**